Update site isolated iframe sizes when they change

achristensen07 opened the pull request:
#### 50e575d266dd7fb9c8e4fe597b2ac4ae1712b5f3
<pre>
Update site isolated iframe sizes when they change
<a href="https://bugs.webkit.org/show_bug.cgi?id=252020">https://bugs.webkit.org/show_bug.cgi?id=252020</a>
rdar://105244387

Reviewed by Simon Fraser.

In order to draw iframes the right size, we get the layer size from the parent
process and send it to the iframe process in a DrawingArea::UpdateGeometry message.

In order to send the DrawingArea::DisplayDidRefresh message to the right process
as a result of a RemoteLayerTreeDrawingAreaProxy::CommitLayerTree message,
instead of calling didRefreshDisplay which does several things that don&apos;t relate
to subframe processes, just send the DrawingArea::DisplayDidRefresh which is the
only part we need for drawing to continue.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
(WebKit::RemoteLayerTreeHost::hostingLayersInverse const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::layerWillBeRemoved):
(WebKit::RemoteLayerTreeHost::createLayer):
* Source/WebKit/UIProcess/SubframePageProxy.cpp:
(WebKit::SubframePageProxy::SubframePageProxy):
(WebKit::SubframePageProxy::~SubframePageProxy):
* Source/WebKit/UIProcess/SubframePageProxy.h:
(WebKit::SubframePageProxy::layerHostingContextIdentifier const):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::commitProvisionalFrame):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::subframePage const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::addHostedFrame):
(WebKit::WebPageProxy::removeHostedFrame):
(WebKit::WebPageProxy::getHostedFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/260406@main">https://commits.webkit.org/260406@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3a9fba86fb104e9e5d6c8ec7f12fe6662b0d439

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108240 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17330 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/41107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/117362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112127 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8609 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100453 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114009 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/18719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/41107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/18719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/41107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/10178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/41107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/10914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/16327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/41107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12497 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3914 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
```diff
@@ -56,7 +56,7 @@ class RemoteLayerTreeDrawingAreaProxy : public DrawingAreaProxy {
     TransactionID lastCommittedLayerTreeTransactionID() const { return m_transactionIDForPendingCACommit; }
 
-    virtual void didRefreshDisplay();
+    virtual void didRefreshDisplay(IPC::Connection*);
```
It's hard for me to tell what this Connection represents.

achristensen07 replied:
Would it be more clear if I changed the name to something like refreshingProcessConnection, somehow changed the parameter to a WebProcessProxy (which works well except will be tricky in the case of RemoteLayerTreeDrawingAreaProxy::commitLayerTree where there currently isn't a mechanism for getting the WebProcessProxy) or somehow wrapped the send(DrawingArea::DisplayDidRefresh) call in a CompletionHandler?

achristensen07 replied:
I'm going with the CompletionHandler approach.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -152,6 +155,19 @@
             m_remoteLayerTreeHost->detachRootLayer();
     }
 
+    for (auto& [layerID, properties] : layerTreeTransaction.changedLayerProperties()) {
```
This seems like an odd place to trigger geometry changes; let's at least factor into `RemoteLayerTreeDrawingAreaProxy::updateSubframeGeometry()`.

Layer bounds is not the right way to get the size of a frame; this data needs to come from the frame tree.

achristensen07 replied:
Could you point me more clearly to where you think this information should come from?  We need to get the size post-layout and this seemed to be the place where it ends up in the hosting layer size, which is what we want to put the iframe contents into.

smfr replied:
This depends on what DrawingAreaProxy–DrawingArea relationship we end up with for subframes. But you should consider the layer tree as "presentation-only"; we only ever push to it, never read back geometry from it.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -339,25 +355,29 @@
 }
 
-void RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay()
+void RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay(IPC::Connection* connection)
```
Again, hard to know what the connection represents. And why doesn't this just fan out to all subframe processes?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -392,7 +412,10 @@
-    m_webPageProxy.sendWithAsyncReply(Messages::DrawingArea::DispatchAfterEnsuringDrawing(), WTFMove(callbackFunction), m_identifier.toUInt64());
+    auto aggregator = CallbackAggregator::create(WTFMove(callbackFunction));
+    m_webPageProxy.sendWithAsyncReply(Messages::DrawingArea::DispatchAfterEnsuringDrawing(), [aggregator] { }, m_identifier);
+    for (auto& process : m_processesWithRegisteredRemoteLayerTreeDrawingAreaProxyMessageReceiver)
```
We'll need to finesse how to drive rendering updates for subprocesses. We want to avoid doing anything to a process for a frame which has nothing to do.

achristensen07 replied:
In this case I'm pretty sure we need to ensure all processes have drawn, but there will be some places where we will want to only do things when there are things to do.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
```diff
@@ -139,6 +139,7 @@
 
     if (auto contextHostID = transaction.remoteContextHostIdentifier()) {
```
We have `contextHostIdentifier`, `contextHostedIdentifier` and `contextHostingIdentifier`. The first one is ambiguous.

smfr commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11944,6 +11944,23 @@ void WebPageProxy::playAllAnimations(CompletionHandler<void()>&& completionHandl
 }
 #endif // ENABLE(ACCESSIBILITY_ANIMATION_CONTROL)
 
+void WebPageProxy::addHostedFrame(WebCore::LayerHostingContextIdentifier layerIdentifier, WebCore::FrameIdentifier frameIdentifier)
```
`layerIdentifier` -> `hostingIdentifier`

smfr commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11944,6 +11944,23 @@ void WebPageProxy::playAllAnimations(CompletionHandler<void()>&& completionHandl
+    m_hostedFrames.set(layerIdentifier, frameIdentifier);
+}
+
+void WebPageProxy::removeHostedFrame(WebCore::LayerHostingContextIdentifier layerIdentifier)
```
Ditto.

smfr commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11944,6 +11944,23 @@ void WebPageProxy::playAllAnimations(CompletionHandler<void()>&& completionHandl
+    m_hostedFrames.remove(layerIdentifier);
+}
+
+WebFrameProxy* WebPageProxy::getHostedFrame(WebCore::LayerHostingContextIdentifier layerIdentifier)
```
Ditto.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -312,6 +312,8 @@
     if (m_viewOverlayRootLayer)
         m_viewOverlayRootLayer->flushCompositingState(visibleRect);
 
+    // FIXME: This siteIsolationEnabled check is a hack because sometimes this gets hit with site isolation.
```
What `siteIsolationEnabled` check?

achristensen07 replied:
Oops, that's from an earlier version of the PR.  Removing...

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -187,8 +196,12 @@
 
-    if (std::exchange(m_didUpdateMessageState, NeedsDidUpdate) == MissedCommit)
-        didRefreshDisplay();
+    if (!layerTreeTransaction.isMainFrameProcessTransaction() || std::exchange(m_didUpdateMessageState, NeedsDidUpdate) == MissedCommit) {
```
This turned something quite simple into something more complicated that hard for me for wrap my head around.

First, let's factor non-main-frame commit stuff into new functions on this class.

Second, I think the completion handler is called synchronously? It would be easier to understand without using a completion handler.

smfr commented on Source/WebCore/page/RemoteFrameClient.h:
```diff
@@ -29,10 +29,13 @@
     WTF_MAKE_FAST_ALLOCATED;
 public:
     virtual void frameDetached() = 0;
+    virtual void sizeUpdated(IntSize) = 0;
```
sizeDidChange?

smfr commented on Source/WebCore/page/RemoteFrameView.cpp:
```diff
@@ -35,6 +36,12 @@ RemoteFrameView::RemoteFrameView(RemoteFrame& frame)
 
+void RemoteFrameView::setFrameRect(const IntRect& rect)
+{
+    m_frame->client().sizeUpdated(rect.size());
```
Can we early return if no change, and call this `sizeDidChange`?

smfr commented on Source/WebCore/page/RemoteFrameView.h:
```diff
@@ -66,6 +66,8 @@ class RemoteFrameView final : public AbstractFrameView {
     void updateLayerPositionsAfterScrolling() final;
     void updateCompositingLayersAfterScrolling() final;
 
+    void setFrameRect(const IntRect&) final;
```
What is the coordinate system of the rect?

achristensen07 replied:
I don't know.

achristensen07 committed:
Update site isolated iframe sizes when they change
https://bugs.webkit.org/show_bug.cgi?id=252020
rdar://105244387

Reviewed by Simon Fraser.

In order to draw iframes the right size, we get the layer size from the parent
process and send it to the iframe process in a DrawingArea::UpdateGeometry message.

In order to send the DrawingArea::DisplayDidRefresh message to the right process
as a result of a RemoteLayerTreeDrawingAreaProxy::CommitLayerTree message,
instead of calling didRefreshDisplay which does several things that don't relate
to subframe processes, just send the DrawingArea::DisplayDidRefresh which is the
only part we need for drawing to continue.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
(WebKit::RemoteLayerTreeHost::hostingLayersInverse const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::layerWillBeRemoved):
(WebKit::RemoteLayerTreeHost::createLayer):
* Source/WebKit/UIProcess/SubframePageProxy.cpp:
(WebKit::SubframePageProxy::SubframePageProxy):
(WebKit::SubframePageProxy::~SubframePageProxy):
* Source/WebKit/UIProcess/SubframePageProxy.h:
(WebKit::SubframePageProxy::layerHostingContextIdentifier const):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::commitProvisionalFrame):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::subframePage const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::addHostedFrame):
(WebKit::WebPageProxy::removeHostedFrame):
(WebKit::WebPageProxy::getHostedFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/260406@main
